common/src/main/java/edu/snu/nemo/common/coder/KVEncoderFactory.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2018 Seoul National University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.nemo.common.coder;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * A kv encoder factory object which generates kv encoders that encode values of type {@code T} into byte streams.
+ * To avoid to generate instance-based coder such as Spark serializer for every encoding,
+ * user need to explicitly instantiate an kv encoder instance and use it.
+ *
+ * @param <T> element type.
+ */
+public interface KVEncoderFactory<T> extends EncoderFactory {
```

Comment:
I feel this adds a lot of boilerplate code.

My understanding this that this PR takes this approach
```
beamKvCoder -> new NemoKVEncoderFactory(beamKvCoder.getKeyCoder(), beamKvCoder.getValueCoder())
```

such that in the IR we can access the key and value coders while bypassing Beam-specific code.

What do you think about this approach? This also bypasses Beam-specific code, perhaps with less code.

```
NemoKeyCoderExtractor -> beamKvCoder.getKeyCoder() OR rddTupleCoder.getKeyCoder()
```